### PR TITLE
Improve AlignmentsTrack by displaying both SNPCoverageTrack and PileupTrack (old alignmentsTrack) when selected

### PR DESCRIPTION
**Pull request by peterkxie**
Addresses step one in issue #792 by combining the old alignments track and snp coverage track into a single track type with subtracks. 

Old alignments track has been renamed to pileup track, and new track is named alignments track and renders both the Pileup track and SNP Coverage track when selected. Config is compatible with any old configs, no changes needed and it will display both tracks when selected from track selector. 

Adding a track allows to select between alignments track or just pileup/snp coverage track individually, but alignments is the default guess.

### CODE REVIEW

**Comment by cmdcolin** [comment]
Is there any plan for how to deal with breakpoint split view with an AlignmentsTrack or will it only work with PileupTrack

**Comment by cmdcolin** [comment]
It's maybe ok for now to have it Pileup only but we may need to see how we can make it work with the new alignments track sometime..notably there may be some changes in how breakpoint split view rendering works a la #820

**Comment by cmdcolin** [comment]
Given that maybe we should just merge this?

**Comment by peterkxie** [comment]
@cmdcolin I think that after the breakpoint split view rendering changes are finalized I can look into making it work with the new alignments track rather than Pileup only

File: packages/linear-genome-view/src/BasicTrack/components/BlockBasedTrack.js [hunk 1]
```diff
@@ -1,17 +1,24 @@
...
 import TrackBlocks from './TrackBlocks'
+import AlignmentsTrackBlocks from './AlignmentsTrackBlocks'
 
 function BlockBasedTrack(props) {
   const { model, children } = props
   return (
     <Track {...props} trackId={getConf(model, 'trackId')}>
       {model.trackMessageComponent ? (
         <model.trackMessageComponent model={model} />
+      ) : getConf(model, 'type') === 'AlignmentsTrack' ? (
```
Comment (rbuels) [inline packages/linear-genome-view/src/BasicTrack/components/BlockBasedTrack.js]: Can't detect AlignmentsTrack here, need to find another way of doing this.  Probably just make a special react component for this track.

File: packages/data-management/src/AddTrackDrawerWidget/components/ConfirmTrack.js [hunk 2]
```diff
@@ -139,10 +139,6 @@ function ConfirmTrack({
               <code>{trackAdapter.type}</code> with subadapter{' '}
               <code>{trackAdapter.subadapter.type}</code>. Please enter a track
               name and, if necessary, update the track type.
-              {/* Using adapter <code>{trackAdapter.type}</code> with subadapter{' '}
```
Comment (rbuels) [inline packages/data-management/src/AddTrackDrawerWidget/components/ConfirmTrack.js]: why is this removed?

File: packages/linear-genome-view/src/BasicTrack/components/BlockBasedTrack.js (hunk 1 above)
Comment (peterkxie) [reply packages/linear-genome-view/src/BasicTrack/components/BlockBasedTrack.js]: special react component made inside the AlignmentsTrack -> component folder. Now AlignmentsTrack returns <AlignmentsBlockBasedTrack/> rather than the BasicTrack's <BlockBasedTrack/>

File: packages/alignments/src/SNPCoverageTrack/configSchema.js [hunk 3]
```diff
@@ -58,7 +49,7 @@ export default pluginManager => {
 
       defaultRendering: {
         type: 'stringEnum',
-        model: types.enumeration('Rendering', ['snpcoverage']),
+        model: types.enumeration('Rendering', ['snpcoverage', 'svg']),
```
Comment (cmdcolin) [inline packages/alignments/src/SNPCoverageTrack/configSchema.js]: what is svg here

File: packages/alignments/src/SNPCoverageAdapter/SNPCoverageAdapter.ts [hunk 4]
```diff
@@ -172,7 +174,7 @@ export default class extends BaseAdapter {
    */
 
   getFeatures(region: IRegion, opts: BaseOptions = {}) {
-    return ObservableCreate<Feature>(async observer => {
+    return ObservableCreate<Feature>(async (observer: Observer<Feature>) => {
```
Comment (cmdcolin) [inline packages/alignments/src/SNPCoverageAdapter/SNPCoverageAdapter.ts]: shouldn't have to manually specify the type of the argument here

File: packages/alignments/src/SNPCoverageAdapter/SNPCoverageAdapter.ts [hunk 5]
```diff
@@ -98,6 +99,7 @@ export default class extends BaseAdapter {
             basesPerSpan: bpPerPx,
           },
         )
+        // @ts-ignore
```
Comment (cmdcolin) [inline packages/alignments/src/SNPCoverageAdapter/SNPCoverageAdapter.ts]: why is ts-ignore needed?

File: packages/jbrowse-web/src/JBrowse.test.js [hunk 6]
```diff
@@ -392,7 +394,49 @@ describe('test configuration editor', () => {
...
+      getAllByTestId('prerendered_canvas_SNPCoverageRenderer'),
+    )
+    // snpCov image
+    const snpCovImg = snpCovCanvas[0].toDataURL()
+    const snpCovData = snpCovImg.replace(/^data:image\/\w+;base64,/, '')
+    const snpCovBuf = Buffer.from(snpCovData, 'base64')
+    expect(snpCovBuf).toMatchImageSnapshot({
+      failureThreshold: 0.5,
+      failureThresholdType: 'percent',
+    })
```
Comment (cmdcolin) [inline packages/jbrowse-web/src/JBrowse.test.js]: nice work here, glad we can get both tested with an image snapshot :+1:

File: packages/linear-genome-view/src/BasicTrack/util/serverSideRenderedBlock.ts [hunk 7]
```diff
@@ -203,7 +204,13 @@ function renderBlockData(self: Instance<BlockStateModel>) {
       sequenceConfig = getSnapshot(trackAssemblyData.sequence.adapter)
     }
     const adapterConfig = getConf(track, 'adapter')
-    const adapterConfigId = jsonStableStringify(adapterConfig)
+    // Only subtracks will have parent tracks with configs
```
Comment (cmdcolin) [inline packages/linear-genome-view/src/BasicTrack/util/serverSideRenderedBlock.ts]: this does assume that the parent track has it's own adapter. should be ok for now

File: packages/wiggle/src/__snapshots__/WiggleRendering.test.js.snap [hunk 8]
```diff
@@ -7,7 +7,7 @@ exports[`one 1`] = `
   style="overflow: visible; position: relative; height: 500px;"
 >
   <canvas
-    data-testid="prerendered_canvas"
+    data-testid="prerendered_canvas_undefined"
```
Comment (cmdcolin) [inline packages/wiggle/src/__snapshots__/WiggleRendering.test.js.snap]: could be worth checking if the undefined here is unintended

File: packages/wiggle/src/WiggleTrack/model.ts [hunk 9]
```diff
@@ -189,7 +188,9 @@ const stateModelFactory = (configSchema: any) =>
                   const aborter = new AbortController()
                   self.setLoading(aborter)
                   const stats = await getStats(aborter.signal)
-                  if (isAlive(self)) {
+
+                  // check if stats returned from empty region
+                  if (isAlive(self) && !stats.emptyRegion) {
```
Comment (cmdcolin) [inline packages/wiggle/src/WiggleTrack/model.ts]: I feel concerned about the stats code having to be concerned about empty region. I feel like maybe a way to approach it is by making it something relevant to the track's logic. Example https://github.com/GMOD/jbrowse-components/compare/792_multiple_renderers_per_track...792_multiple_renderers_per_track_suggestion

I know that is picky but it seems like the stats code should be more or less just about the data whereas this empty region weirdness is really more about the track being setup

File: packages/wiggle/src/__snapshots__/WiggleRendering.test.js.snap (hunk 8 above)
Comment (peterkxie) [reply packages/wiggle/src/__snapshots__/WiggleRendering.test.js.snap]: The WiggleRendering.test and PileupRendering.test had empty config objects which lead to undefined type. Since I newly appended config.type to the testid, config was not needed to pass in previously for the test. I have now added a config object with a type: DummyRenderer to the test call so undefined should no longer appear. Does not look like an issue outside of test files, have not seen any surprise 'undefined' with the config.type

File: packages/alignments/src/SNPCoverageAdapter/SNPCoverageAdapter.ts (hunk 5 above)
Comment (peterkxie) [reply packages/alignments/src/SNPCoverageAdapter/SNPCoverageAdapter.ts]: left in from when scoresToStats expected an Observable<BBIFeature> for feats, scoresToStats has since been changed but forgot to remove this and will be removed in next commit

File: packages/wiggle/src/WiggleTrack/model.ts (hunk 9 above)
Comment (peterkxie) [reply packages/wiggle/src/WiggleTrack/model.ts]: Valid concern and I agree. Thanks for the suggestion code, have implemented and tested around. Addresses the issue of premature empty updating of stats without having to check for a specific stats object returned

File: packages/alignments/src/SNPCoverageTrack/configSchema.js (hunk 3 above)
Comment (peterkxie) [reply packages/alignments/src/SNPCoverageTrack/configSchema.js]: Svg was included because PileupTrack could be chosen between pileup and svg defaultRendering and since they share the same parent config editor svg would be applied to SNPCoverageTrack which would error. 

Better solution that will be implemented is to remove defaultRendering field from SNPCoverageTrack since it only has one renderer anyways.

This may need addressing in future if subtrack concept is expanded upon and the subtracks each have multiple renderers to select from and do not overlap. Will probably have to change the defaultRendering name to be subtrack specific aka defaultPileupRendering or similar idea
